sky/serve/autoscalers.py

```diff
@@ -132,93 +130,100 @@ def collect_request_information(
-
+        target_num_replicas = math.ceil(num_requests_per_second /
+                                        self.target_qps_per_replica)
         target_num_replicas = max(self.min_replicas,
                                   min(self.max_replicas, target_num_replicas))
+        logger.info(f'Requests per second: {num_requests_per_second}, '
+                    f'Current target number of replicas: {target_num_replicas}')
 
-        logger.info(f'Target number of replicas: {target_num_replicas}, '
-                    f'min_replicas: {self.min_replicas}, '
-                    f'max_replicas: {self.max_replicas}')
-
-        num_replicas_delta = 0
-        if num_replicas < self.min_replicas:
-            num_replicas_delta = self.min_replicas - num_replicas
-            self.upscale_counter = 0
-        elif num_replicas > self.max_replicas:
-            num_replicas_delta = self.max_replicas - num_replicas
-            self.downscale_counter = 0
-        elif target_num_replicas > num_replicas:
+        if target_num_replicas > self.target_num_replicas:
             self.upscale_counter += 1
             self.downscale_counter = 0
             if self.upscale_counter >= self.scale_up_consecutive_periods:
                 self.upscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
-        elif target_num_replicas < num_replicas:
+                return target_num_replicas
+        elif target_num_replicas < self.target_num_replicas:
             self.downscale_counter += 1
             self.upscale_counter = 0
             if self.downscale_counter >= self.scale_down_consecutive_periods:
                 self.downscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
+                return target_num_replicas
         else:
             self.upscale_counter = self.downscale_counter = 0
-        logger.info(f'Upscale counter: {self.upscale_counter}/'
-                    f'{self.scale_up_consecutive_periods}. '
-                    f'Downscale counter: {self.downscale_counter}/'
-                    f'{self.scale_down_consecutive_periods}')
+        return self.target_num_replicas
 
-        if num_replicas_delta == 0:
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+
+        # Don't count over-provision here.
+        self.target_num_replicas = self._get_desired_num_replicas()
+        logger.info(
+            f'Final target number of replicas: {self.target_num_replicas} '
+            f'({self.target_num_replicas} with '
+            f'over-provision), Upscale counter: {self.upscale_counter}/'
+            f'{self.scale_up_consecutive_periods}, '
+            f'Downscale counter: {self.downscale_counter}/'
+            f'{self.scale_down_consecutive_periods}')
+
+        alive_replica_infos = [info for info in replica_infos if info.is_alive]
```

Comment:
It seems intuitive that we should consider only alive replicas when deciding autoscaling? since failed replicas should not affect autoscaling decisions?

Reply:
We also have `auto_restart=False` settings, which should count the failed replica as a replica and prevent infinite restart for error user code.

Reply:
hmm it seems we can just always exclude failed replicas (even with `auto_restart=False`)?

Reply:
or we can put an assert for auto_restart